vault/resource_kv_secret_test.go

```diff
@@ -69,3 +75,46 @@ resource "vault_kv_secret" "test" {
 
 	return ret
 }
+
+func testKVSecretConfigUpdate(mount, name string) string {
+	ret := fmt.Sprintf(`
+%s
+
+`, kvV1MountConfig(mount))
+
+	ret += fmt.Sprintf(`
+resource "vault_kv_secret" "test" {
+  path = "${vault_mount.kvv1.path}/%s"
+  data_json = jsonencode(
+    {
+      zip = "zoop",
+      foo = "bar"
+    }
+  )
+}`, name)
+
+	return ret
+}
+
+func testResourceKVSecret_updateCheck(s *terraform.State) error {
```

Comment:
We prefer using the `resource.ComposeTestCheckFunc` test code style whenever possible, as done in L27; something like 

```
resource.ComposeTestCheckFunc(
    resource.TestCheckResourceAttr(resourceName, consts.FieldPath, fmt.Sprintf("%s/%s", mount, name)),
    resource.TestCheckResourceAttr(resourceName, "data.zip", "zoop"),					 
    resource.TestCheckResourceAttr(resourceName, "data.foo", "bar"),
),
```

Reply:
With the suggested implementation of the ComposeTestCheckFunc the test with the previous implementation in resource_kv_secret.go will not fail, because the extra "data" in the path is implemented in kvSecretWrite as well as in kvSecretRead.

This is why I made the test fail with the first commit and fixed the implementation with the second commit. I had to use the api client directly to make the test fail, because the read context compensated the failure in the create context. It is somehow that the terraform provider is not a tool for itself but for configuring the application to be used (here for retrieving values by api or cli).

Reply:
Since we have the Import test on L38, we are already take care of checking the TF config values against what is configured in Vault. The composition of test funcs would also make comparing the two maps simpler, as opposed to only checking a single key. I'd still suggest removing this extra check func and adding the precedent style of `ComposeTestCheckFunc`  :)

Reply:
I refactored the update test to composition.

I added an extra test step for the api access check.

I think that the import test also uses the state read by the provider. Thus the test does not fail. You can verify this by checking out https://github.com/l-with/terraform-provider-vault/tree/does-kv-v1-test-fail and run the test. It will not fail. It will fail with the added api access check function (https://github.com/l-with/terraform-provider-vault/blob/kv-v1-test-fail/vault/resource_kv_secret_test.go).

Reply:
> I refactored the update test to composition.

Thanks!

> 
> I added an extra test step for the api access check.
> 
> I think that the import test also uses the state read by the provider. Thus the test does not fail. You can verify this by checking out https://github.com/l-with/terraform-provider-vault/tree/does-kv-v1-test-fail and run the test. It will not fail. It will fail with the added api access check function (https://github.com/l-with/terraform-provider-vault/blob/kv-v1-test-fail/vault/resource_kv_secret_test.go).

The import does a complete read of the resource from Vault into a new state, and verifies that there are no differences between the TF code and that state. So I think we are covered here without the extra direct API check.

Reply:
I have tried it out, only the direct API check fails with the implementation of the current release that nests the kv v1 secrets in extra path component "data". Also the test in the current implementation does not fail, the added update test makes no difference. 

Surely this is a rare case, but since the the current release has the issue, I think it is reasonable that the fix contains the test that reveals the issue together with the fix of the issue.

Reply:
I think I agree with you, we can keep the regression check in this case.